ProcessOptimizer/samplers/doe/optimal_design.py

```diff
@@ -0,0 +1,697 @@
+    Copyright 2016 Stat-Ease, Inc.
+    License: Apache License, Version 2.0
+    License link: https://github.com/statease/dexpy/blob/master/LICENSE
+    """
+    F2 = np.vstack((new_point, old_point))
+    F1 = F2.T.copy()
+    F1[:, 1] *= -1
+    FD = np.dot(F2, XtXi)
+    I2x2 = np.identity(2) + np.dot(FD, F1)
+    Inverse2x2 = scipy.linalg.inv(I2x2)
+    F2x2FD = np.dot(np.dot(F1, Inverse2x2), FD)
+    return XtXi - np.dot(XtXi, F2x2FD)
+
+
+def expand_point(design_point, code):
+    """Converts a point in factor space to conform with the X matrix.
+
+    This function is from https://github.com/statease/dexpy
+    Copyright 2016 Stat-Ease, Inc.
+    License: Apache License, Version 2.0
+    License link: https://github.com/statease/dexpy/blob/master/LICENSE
+    """
+    return np.array(eval(code, {}, design_point))
+
+
+def delta(X, XtXi, row, new_point):
+    """Calculates the change in D-optimality from exchanging a point.
+
+    This is equation (1) in Meyer and Nachtsheim :cite:`MeyerNachtsheim1995`.
+
+    This function is from https://github.com/statease/dexpy
+    Copyright 2016 Stat-Ease, Inc.
+    License: Apache License, Version 2.0
+    License link: https://github.com/statease/dexpy/blob/master/LICENSE
+    """
+    old_point = X[row]
+
+    added_variance = np.dot(new_point, np.dot(XtXi, new_point.T))
+    removed_variance = np.dot(old_point, np.dot(XtXi, old_point.T))
+    covariance = np.dot(new_point, np.dot(XtXi, old_point.T))
+    return (
+        1
+        + (added_variance - removed_variance)
+        + (covariance * covariance - added_variance * removed_variance)
+    )
+
+
+def make_model(factor_names, model_order, include_powers=True):
+    """Creates patsy formula representing a given model order.
+
+    :param factor_names: The names of the factors in the design.
+    :type factor_names: list of str
+
+    :param model_order: The order of the model.
+    :type model_order: int
+
+    :param include_powers: Whether to include squared and cubed terms.
+    :type include_powers: bool, list of bool
+    if True, include squared and cubed terms for all factors
+    if include_powers is a list, it must be the same length as factor_names
+    if include_powers is a list, include squared and cubed terms for factors
+    where include_powers is True
+
+    :return: The patsy formula representing the model.
+    :rtype: str
+
+    This function is inspired by similar function in
+    https://github.com/statease/dexpy
+    Copyright 2016 Stat-Ease, Inc.
+    License: Apache License, Version 2.0
+    License link: https://github.com/statease/dexpy/blob/master/LICENSE
+    """
+
+    if isinstance(include_powers, list) and len(include_powers) != len(
+        factor_names
+    ):
+        raise ValueError(
+            "The length of include_powers must be equal to the number of "
+            "factors"
+        )
+
+    if model_order == 1:
+        return "+".join(factor_names)
+
```

Comment:
```suggestion
if include_powers == True:
    include_powers = [True]*len(factor_names)
elif include_powers = False:
    include_powers = [False]*len(factor_names)
```
That gets rid of the if, if, else in the two cases below. It does take longer to run, but that is probably not significant.

Reply:
Nope. Function could use some cleanup, but it is this way to make sure we do not take power terms of categorical variables